keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -3,111 +3,556 @@
+- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket protected by standard file permissions. Documentation will be added to ensure that distributions limit the socket to privileged or specific uids.
+- With External signer, token headers and claims can diverge from that in existing JWTs as signed by kube-apiserver. To mitigate, external signers will only return signature and header while kube-apiserver will assemble the JWT. This will allow kube-apiserver to validate the header and control the claims.
+
+## Design Details
+
+### ExternalJWTSigner RPC
+
+Will be served on a socket as configured via the `--service-account-signing-endpoint` flag.
 
 ```proto
 syntax = "proto3";
 
-package v1alpha1;
+// This service is served by a process on a local Unix Domain Socket.
+service ExternalJWTSigner {
+  // Sign takes a serialized JWT payload, and returns the serialized header and
+  // signature.  The caller can then assemble the JWT from the header, payload,
+  // and signature.
+  //
+  // The plugin MUST set a key id in the returned JWT header.
+  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}
+
+  // FetchKeys returns the set of public keys that are trusted to sign
+  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
+
+  <<[UNRESOLVED]>>
+  
+  // Status is meant to be polled for external signer's health status.
+  rpc Status(StatusRequest) returns (StatusResponse) {}
+  
+  <<[/UNRESOLVED]>>
+}
 
-service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+message SignJWTRequest {
+  // URL-safe base64 wrapped payload to be signed.
+  // Exactly as it appears in the second segment of the JWT
+  bytes claims = 1;
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
-  bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
+message SignJWTResponse {
+  // header must contain alg, kid, typ claims.
+  // type must be “JWT”.
+  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).
+  // Already wrapped in URL-safe base64, valid for directly inserting into the final JWT.
+  bytes header = 1;
+
+  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  bytes signature = 2;
```

Comment:
Since these fields are base64, should this be a `string`?

Reply:
changed both, header and signature to string values.